test/test_sparse_csr.py

```diff
@@ -3978,28 +3979,44 @@ def test_TensorAsKey(self, device):
         # but key is still valid:
         self.assertEqual(d.get(key5), (key5, 567), **assertEqualOptions)
 
-    @parametrize("op", ['bsr_dense_addmm', 'bsr_dense_mm', 'bsr_dense_linear'])
+    @suppress_warnings
+    @parametrize("op", ['bsr_dense_addmm', 'bsr_dense_mm', 'bsr_dense_linear', '_int_bsr_dense_addmm'])
     @parametrize("blocksize", [16, '16x32', 32])
     @onlyCUDA
     @skipIfRocm
-    @dtypes(torch.half, torch.bfloat16, torch.float)
-    @dtypesIfCUDA(torch.half, *[torch.bfloat16] if SM80OrLater else [], torch.float)
+    @dtypes(torch.half, torch.bfloat16, torch.float, torch.int8)
+    @dtypesIfCUDA(torch.half, *[torch.bfloat16] if SM80OrLater else [], torch.float, torch.int8)
     @unittest.skipIf(IS_FBCODE and IS_REMOTE_GPU, "Test requires Triton")
     def test_triton_kernel(self, op, device, dtype, blocksize):
-        from torch.sparse._triton_ops import bsr_dense_addmm, bsr_dense_mm
+        from torch.sparse._triton_ops import bsr_dense_addmm, bsr_dense_mm, _int_bsr_dense_addmm
         from torch.sparse._triton_ops_meta import (create_blocked_tensor, get_meta,
                                                    optimize_bsr_dense_addmm, dump)
 
         def bsr_dense_linear(input, weights, bias=None):
             return torch.nn.functional.linear(input, weights, bias=bias).transpose(-1, -2)
 
-        operation = dict(bsr_dense_addmm=bsr_dense_addmm, bsr_dense_mm=bsr_dense_mm, bsr_dense_linear=bsr_dense_linear)[op]
+        operation = dict(bsr_dense_addmm=bsr_dense_addmm, bsr_dense_mm=bsr_dense_mm, bsr_dense_linear=bsr_dense_linear,
+                         _int_bsr_dense_addmm=_int_bsr_dense_addmm)[op]
 
-        def reference(input, mat1, mat2, beta=1, alpha=1):
+        def reference(input, mat1, mat2, beta=1, alpha=1, op=op):
             assert mat1.layout is torch.strided
             assert mat2.layout is torch.strided
+            if dtype is torch.int8:
+                if op == '_int_bsr_dense_addmm':
+                    return beta * input + alpha * torch._int_mm(mat1, mat2)
+                # workaround RuntimeError: "addmm_cuda" not implemented for 'Char'
+                return beta * input + alpha * torch._int_mm(mat1, mat2).to(torch.int8)
```

Comment:
It should be possible to just use `_int_mm` now without the downcast

Reply:
yes, this is when one is using `_int_bsr_dense_addmm`, see the two lines above.

When using `bsr_dense_addmm`, the classical convention is used: by default, the dtype of input will be the dtype of output.

Reply:
Hm, but `_int_bsr_dense_addmm` should return int32 when passed int8.

So why do `torch._int_mm(mat1, mat2)`*`.to(torch.int8)`*?

I'd expect `beta * input + alpha * torch._int_mm(mat1, mat2)` to be sufficient and also `torch._int_mm(mat1, mat2).to(torch.int8)` to cause errors when `mat1` and `mat2` are sufficiently large. Maybe `mat1` and `mat2` don't use the full int8 range.

Reply:
> So why do torch._int_mm(mat1, mat2).to(torch.int8)?

See https://github.com/pytorch/pytorch/pull/133855#discussion_r1725699844. This corresponds to feature 1.

> I'd expect beta * input + alpha * torch._int_mm(mat1, mat2) to be sufficient..

Yes, this is indeed used for feature 2 (notice `if op == '_int_bsr_dense_addmm':`).

> also torch._int_mm(mat1, mat2).to(torch.int8) to cause errors when mat1 and mat2 are sufficiently large

I agree. The high risk of int8 overflow is unavoidable within the feature 1.